mostlyai/engine/_common.py

```diff
@@ -610,3 +614,264 @@ def clear(self):
+
+    Args:
+        values (np.ndarray): A 1D array of numeric values.
+        epsilon (float): The privacy budget to spend estimating the bounds.
+
+    Returns:
+        tuple[float, float]: A tuple of the estimated minimum and maximum values.
+    """
+
+    bins = 64
+    hist = [0.0] * bins * 2
+
+    values = np.array(values, dtype=np.float64)
+    values = values[values != np.inf]
+    values = values[values != -np.inf]
+    values = values[~np.isnan(values)]
+
+    def edges(idx):
+        if idx == bins:
+            return (0.0, 1.0)
+        elif idx > bins:
+            return (2.0 ** (idx - bins - 1), 2.0 ** (idx - bins))
+        elif idx == bins - 1:
+            return (-1.0, -0.0)
+        else:
+            return (-1 * 2.0 ** np.abs(bins - idx - 1), -1 * 2.0 ** np.abs(bins - idx - 2))
+
+    edge_list = [edges(idx) for idx in range(len(hist))]
+    min_val = min([lower for lower, _ in edge_list])
+    max_val = max([upper for _, upper in edge_list]) - 1
+    values = np.clip(values, min_val, max_val)
+
+    # compute histograms
+    for v in values:
+        bin = None
+        for idx, (lower, upper) in enumerate(edge_list):
+            if lower <= v < upper:
+                bin = idx
+                break
+        if bin is None:
+            bin = idx
+        hist[bin] += 1
+
+        # for testing
+        lower, upper = edges(bin)
+
+    n_bins = len(hist)
+
+    noise = np.random.laplace(loc=0.0, scale=1 / epsilon, size=n_bins)
+    hist = [val + lap_noise for val, lap_noise in zip(hist, noise)]
+
+    failure_prob = 10e-9
+    highest_failure_prob = 1 / (n_bins * 2)
+
+    exceeds = []
+    while len(exceeds) < 1 and failure_prob <= highest_failure_prob:
+        p = 1 - failure_prob
+        K = -np.log(2 - 2 * p ** (1 / (n_bins - 1))) / epsilon
+        exceeds = [idx for idx, v in enumerate(hist) if v > K]
+        failure_prob *= 10
+
+    if len(exceeds) == 0:
+        return (None, None)
+
+    lower_bin, upper_bin = min(exceeds), max(exceeds)
+    lower, _ = edges(lower_bin)
+    _, upper = edges(upper_bin)
+    return (float(lower), float(upper))
+
+
+def _dp_bounded_quantiles(
+    values: np.ndarray, quantiles: list[float], epsilon: float, lower: float, upper: float
+) -> list[float]:
+    """
+    Estimate the quantile.
+    from: http://cs-people.bu.edu/ads22/pubs/2011/stoc194-smith.pdf
+
+    Args:
+        values (np.ndarray): A 1D array of numeric values.
+        quantiles (list[float]): List of probabilities of the quantiles to estimate.
+        epsilon (float): Privacy budget.
+        lower (float): A bounding parameter. The quantile will be estimated only for values greater than or equal to this bound.
+        upper (float): A bounding parameter. The quantile will be estimated only for values less than or equal to this bound.
+
+    Returns:
+        list[float]: The estimated quantile.
+    """
+
+    _LOG.info(f"compute DP bounded quantiles within [{lower}, {upper}]")
```

Comment:
would it make sense to reorder those functions in file?
`_dp_approx_bounds`, `_dp_bounded_quantiles`, `_dp_unbounded_quantiles`?